Bluetooth: controller: fixup HCI LE Create CIS 'parameter' checks

erbr-ot opened the pull request:
A couple of minor updates to condition checks for HCE LE Create CIS:
* BT Core Spec states that feature exchange should be initiated if not already performed.
* Also confirm that the 'requesting connection' is acting as central

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_llcp.c:
```diff
@@ -874,8 +874,11 @@ uint8_t ull_cp_cis_create(struct ll_conn *conn, struct ll_conn_iso_stream *cis)
 
-	if (conn->lll.handle != cis->lll.acl_handle) {
-		return BT_HCI_ERR_CMD_DISALLOWED;
+	if (!conn->llcp.fex.valid) {
+		/* No feature exchange was performed so initiate before CIS Create */
+		if (ull_cp_feature_exchange(conn) != BT_HCI_ERR_SUCCESS) {
+			return BT_HCI_ERR_CMD_DISALLOWED;
+		}
```
- Do we want to do auto feature exchange?
- How is controller initiated control procedure differentiated from Host initiated?
- Where is the CIS support checked once feature exchange completes?
- How and what event to be returned if remote does not support CIS?

thoh-ot replied:
It also ties well into

> 4.5.11 Control PDU length management
> The Link Layer shall not transmit a packet containing an LL Control PDU with a
> CtrData field longer than 26 octets until it has successfully completed a Feature
> Exchange procedure (see Section 5.1.4) on the same connection.
>
> If the Link Layer in the Central role supports receiving LL Control PDUs with a
> CtrData field longer than 26 octets, it should initiate the Feature Exchange
> procedure on each connection.
>
> Note: As specified in Section 4.6, once the feature exchange has completed,
> the Link Layer must not use a procedure that the peer did not mark as
> supported. Therefore the Link Layer will never transmit an LL Control PDU with
> a CtrData field longer than 26 octets to a device that does not support it.

And LL_CIS_REQ is 35 bytes.

erbr-ot replied:
Re. 1: I guess they are not. I think that should not be a problem. As far as I understand - an extra FEAT_EXCH could then be performed, but that ought not be a problem.
Re. 2: It is not, and that is a problem! A solution could be to implement at check for feature support in the CIS Create procedure (before sending CIS_REQ) and complete with UNSUPPORTED in case it is? Then if a CIS create is enqueued 'pending' completion of FEAT_EXCH would be aborted, signaling 
Re. 3: see above?

cvinayak replied:
Changes implemented.

erbr-ot committed:
Bluetooth: controller: fixup HCI LE Create CIS 'parameter' checks

BT Core Spec states that feature exchange should be initiated if
not already performed. So in this case initiate feature exchange
and 'abort' enqueued CIS create if peer doesn't support CIS peripheral
feature

Signed-off-by: Erik Brockhoff <erbr@oticon.com>

erbr-ot committed:
Bluetooth: controller: confirm central role on CIS Create

Confirm that the 'requesting connection' is acting as central

Signed-off-by: Erik Brockhoff <erbr@oticon.com>

erbr-ot committed:
Bluetooth: controller: Only generate ntf on host initiated feature exch

Only generate notification when feature exchange was initiated by host

Signed-off-by: Erik Brockhoff <erbr@oticon.com>

cvinayak commented:
Fix for the bsim test failure: https://github.com/zephyrproject-rtos/zephyr/pull/57760

kruithofa reviewed (approved):
LGTM

te-johan commented on subsys/bluetooth/controller/ll_sw/ull_llcp_cc.c:
```diff
@@ -838,13 +838,33 @@ static void lp_cc_st_wait_tx_cis_req(struct ll_conn *conn, struct proc_ctx *ctx,
+			/* In case feature exchange completed after CIS create was enqueued
+			 * peer CIS peripheral support should be confirmed
+			 */
+			if (feature_peer_iso_peripheral(conn)) {
```
this call seems to break things since feature_peer_iso_peripheral checks if LL_FEAT_BIT_CIS_PERIPHERAL is set.
LL_FEAT_BIT_CIS_PERIPHERAL is set to zero if the controller does not support cis peripheral.
in my case my controller only supports cis central.

the actual issue lies in the feature check function. they should use the BIT64(BT_LE...) defines instead.

erbr-ot replied:
Correct - should be fixed by: https://github.com/zephyrproject-rtos/zephyr/pull/57862

te-johan replied:
thanks for the quick reply. that might solve other issues as well.
